### PR TITLE
MINIFICPP-1959 - Avoid dangling content references when using VolatileFlowFileRepository

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/test/unit/ProcessSessionTests.cpp [hunk 1]
```diff
@@ -126,3 +132,44 @@ TEST_CASE("ProcessSession::read can read zero length flowfiles without crash", "
...
   ContentRepositoryDependentTests::testReadFromZeroLengthFlowFile(std::make_shared<core::repository::FileSystemRepository>());
 }
+
+struct VolatileFlowFileRepositoryTestAccessor {
+  METHOD_ACCESSOR(flush);
+};
+
+class TestVolatileFlowFileRepository : public core::repository::VolatileFlowFileRepository {
+ public:
+  explicit TestVolatileFlowFileRepository(std::string name) : core::SerializableComponent(std::move(name)) {}
```
Comment (fgerlits) [inline libminifi/test/unit/ProcessSessionTests.cpp:142]: why is this not
```suggestion
  explicit TestVolatileFlowFileRepository(std::string name) : core::repository::VolatileFlowFileRepository(std::move(name)) {}
```
?

Comment (adamdebreceni) [reply libminifi/test/unit/ProcessSessionTests.cpp:142]: `SerializableComponent` is a virtual base class without default initializer, so we have to directly initialize it in every derived class

File: libminifi/include/core/repository/VolatileRepository.h [hunk 2]
```diff
@@ -256,7 +257,7 @@ bool VolatileRepository<KeyType, RepositoryType>::Put(const KeyType& key, const
   } while (!updated);
   current_size_ += size;
 
-  logger_->log_debug("VolatileRepository -- put %u %u", current_size_.load(), current_index_.load());
+  logger_->log_debug("VolatileRepository -- put %zu " PRIu32, current_size_.load(), current_index_.load());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/repository/VolatileRepository.h]: added

**Review by fgerlits — APPROVED** [review]

File: libminifi/test/unit/ProcessSessionTests.cpp [hunk 3]
```diff
@@ -138,3 +144,44 @@ TEST_CASE("ProcessSession::read can read zero length flowfiles without crash", "
...
+  const auto flow_file_2 = session.create();
+  const auto flow_file_3 = session.create();
+  session.transfer(flow_file_1, Success);
+  session.transfer(flow_file_2, Success);
+  session.transfer(flow_file_3, Success);
+  session.commit();
+
+  REQUIRE(flow_file_1->getResourceClaim()->getFlowFileRecordOwnedCount() == 1);
+  REQUIRE(flow_file_2->getResourceClaim()->getFlowFileRecordOwnedCount() == 2);
+  REQUIRE(flow_file_3->getResourceClaim()->getFlowFileRecordOwnedCount() == 2);
```
Comment (adam-markovics) [inline libminifi/test/unit/ProcessSessionTests.cpp:182]: Why is it 1, 2, 2? I thought it should be 0, 1, 1. Who is the other owner than `ff_repo`?

File: libminifi/src/core/ProcessSession.cpp [hunk 4]
```diff
@@ -958,54 +958,72 @@ void ProcessSession::persistFlowFilesBeforeTransfer(
...
+        }
       }
+    }
+  };
 
-      std::unique_ptr<io::BufferStream> stream(new io::BufferStream());
-      std::static_pointer_cast<FlowFileRecord>(ff)->Serialize(*stream);
+  // collect serialized flowfiles
+  forEachFlowFile(Type::Transferred, [&] (auto& ff, auto& /*original*/) {
+    std::unique_ptr<io::BufferStream> stream(new io::BufferStream());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/ProcessSession.cpp]: replaced

File: libminifi/test/TestBase.h [hunk 5]
```diff
@@ -331,10 +331,19 @@ class TestPlan {
 
 class TestController {
  public:
+  struct PlanConfig {
+    std::shared_ptr<minifi::Configure> configuration = {};
+    const char* state_dir = nullptr;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/test/TestBase.h]: changed it to `std::optional<std::filesystem::path>`

File: libminifi/test/unit/ProcessSessionTests.cpp (hunk 3 above)
Comment (adamdebreceni) [reply libminifi/test/unit/ProcessSessionTests.cpp:184]: added comment, the shared_ptr on the stack is the other owner

**Review by lordgamez — APPROVED** [review]

File: libminifi/test/unit/ProcessSessionTests.cpp (hunk 3 above)
Comment (adam-markovics) [reply libminifi/test/unit/ProcessSessionTests.cpp:184]: Thanks, it's more clear now!

**Review by adam-markovics — APPROVED** [review]

**Pull request closed without merging**